### PR TITLE
feat(View): add support for styleUrls

### PR DESCRIPTION
**Pull request by vicb**
@mhevery When `styles: []` are specified on the `@View`, they will be handled by the shadow DOM strategy (`@import` are inlined, selector are re-written).

Ultimately when `styles` will be populated during the build phase, there will be no need to transform the styles (already transformed).
- Is this ok to transform them for now or should we already switch to the end strategy ?
- Do we really need `styles: List` or `style: string` is ok ? (concatenation could also happen at build time) ? 

TODO:
- [ ] always inline (#1694) -> SD strategies become sync
- [ ] `TemplateLoader` -> `ViewLoader`

### CODE REVIEW

File: modules/angular2/src/render/dom/compiler/compiler.ts [hunk 1]
```diff
@@ -73,8 +74,52 @@ export class DomCompiler extends RenderCompiler {
...
+    }
+
+    // Fastest case when the view has no inline nor external styles.
+    return pvdPromise;
+  }
+
+  private _injectCss(componentId: string, tplEl, urls: string | List<string>,
+                     styles: List<string>) {
+    for (var i = styles.length - 1; i >= 0; i--) {
+      let url = urls instanceof String ? urls : urls[i];
```
Comment (gdi2290) [inline modules/angular2/src/render/dom/compiler/compiler.ts]: `let i` and facade isString

File: modules/angular2/src/render/dom/compiler/compiler.ts [hunk 2]
```diff
@@ -73,8 +74,52 @@ export class DomCompiler extends RenderCompiler {
 
 @Injectable()
 export class DefaultDomCompiler extends DomCompiler {
-  constructor(parser: Parser, shadowDomStrategy: ShadowDomStrategy,
+  constructor(parser: Parser, public shadowDomStrategy: ShadowDomStrategy,
               templateLoader: TemplateLoader) {
     super(new DefaultStepFactory(parser, shadowDomStrategy), templateLoader);
   }
+
+  compile(template: ViewDefinition): Promise<ProtoViewDto> {
```
Comment (tbosch) [inline modules/angular2/src/render/dom/compiler/compiler.ts]: Why is this in `DefaultCompiler` and not in `DomCompiler`? The purpose of the `DefaultDomCompiler` is only to provide default steps, but not to add new functionality...

**Comment by tbosch** [comment]
I think there are 2 independent things here:

1. load and inline styles: they can come either from `styles`/`stylesUrl` annotation or be embedded in the template
2. adjust the styles for shadow dom emulation

1) needs to be done independently of the shadow dom emulation: We don't inline for native shadow DOM right now, but we should, as otherwise view instantiation gets asynchronous (we would need to wait for the `<style href="...">` elements have been loaded (see #1694). This needs to happen before any compile step. I would vote for letting the `TemplateLoader` do this already, so a build step that inlines the template would automatically also inline the styles.

2) is depending on the shadow dom emulation emulation and can assume that styles are already inline in the template as a `<style>` element. I.e. this can be a regular compile step.

A side effect of this would be that the `CompileStepFactory` no longer needs `subTaskPromises`, as they were only used by the compile step that inlines styles. Also, `ShadowDomStrategy.processStyleElement` would always be synchronous as it would not do any loading.

@vicb Could you change the code accordingly?

File: modules/angular2/src/render/dom/compiler/compiler.ts (hunk 2 above)
Comment (vicb) [inline modules/angular2/src/render/dom/compiler/compiler.ts]: because we need to Strategy to inject the CSS

**Comment by vicb** [comment]
1) 
> This needs to happen before any compile step

Why ?
I think we can parallelize compilation & css loading / inlining, what would be wrong with that ?
 
> TemplateLoader

Should we rename it to `ViewLoader` now and have `loadTemplate` and `loadStyles` methods ?

2) Not sure to get why we need to delay the compilation ? Could this be updated in a second step.

May be we should chat today to clarify this ?

**Comment by tbosch** [comment]
talked to @mhevery: Please inline first and then start the compile process. This is ok for production as in that case we would inline the html and styles in a build step so there would be no xhrs.

**Comment by vicb** [comment]
Thanks for the feedback, I'll update tomorrow.

**Comment by vicb** [comment]
@tbosch could you please review this PR ?

Currently I have implemented what we've discussed yesterday: move the styles from the annotation inside the template before the compilation.

In order to keep the PR reviewable, I'll address the remaining stuff (see TODO in the description above) as separate PRs.

File: modules/angular2/src/render/dom/compiler/template_loader.ts [hunk 3]
```diff
@@ -15,39 +15,62 @@ import {UrlResolver} from 'angular2/src/services/url_resolver';
...
+
+          return DOM.createTemplate(html);
+        });
   }
+
+  private _loadText(url: string): Promise<string> {
+    var response = MapWrapper.get(this._cache, url);
+
+    if (isBlank(response)) {
+      // TODO(vicb): change error when TS gets fixed
```
Comment (tbosch) [inline modules/angular2/src/render/dom/compiler/template_loader.ts]: Maybe 'TODO: remove `PromiseWrapper.reject` and throw directly'?

Comment (tbosch) [inline modules/angular2/src/render/dom/compiler/template_loader.ts]: Never mind...

**Comment by tbosch** [comment]
Looks good. Please add link to the PR that will make ShadowDom strategy sync.
And please add `fixes #1694` to the last commit as well.

**Comment by tbosch** [comment]
Needs rebase though...

**Comment by vicb** [comment]
@tbosch #1694 will be fixed as part of the next PR (strategies are still async in this one)

**Comment by vicb** [comment]
landed as ac3e624d0fbb8c67f1df3a928cc23d70d3ad5542 + previous commits

**Pull request closed without merging**